src/ck-core/ck.C

```diff
@@ -1705,7 +1709,11 @@ static inline void _sendMsgBranch(int eIdx, void *msg, CkGroupID gID,
         env = _prepareImmediateMsgBranch(eIdx,msg,gID,ForBocMsg);
     }else
     {
-        env = _prepareMsgBranch(eIdx,msg,gID,ForBocMsg);
+        if (pe == CLD_BROADCAST || pe == CLD_BROADCAST_ALL) {
```

Comment:
@epmikida: Is there a specific reason for not guarding the `BocBcastMsg` branch with `#if CMK_SMP`, since there's no reason for non-smp group messages to be marked `BocBcastMsg`. I think for regular messages it doesn't matter since both the delivery routes end up working fine, although `ForBocMsg` maybe be a faster route. 

However for within node  + ZC API, in the nonSMP mode, I see that ZC Bcast Post API for groups crashes with `Assertion "startNode >=0 && startNode<_Cmi_numnodes" failed in file machine-broadcast.C` on intermediate nodes because of the message marked `BocBcastMsg`. This happens because of the message going through the scheduler twice (as it happens in ZC Bcast Post API) and the bcast root being set again on the message. When I added the `CMK_SMP` guard, it seems to work fine.

Reply:
Ok that sounds good. I think a careful look at the need for CMK_SMP would be useful for Charm 7, but for now I think it makes sense here.

Reply:
Okay, I will add this guard when I'm adding the new minor changes that are now required for getting the Zerocopy Bcast API working.